File: src/bin/day7.rs
const INPUT: &str = include_str!("inputs/day7.txt");

fn runops( nums: &[usize], inx: usize, target: usize, acc: usize, do2: bool) -> bool {
    if acc > target {
        return false;
    }
    if inx == nums.len() {
        return acc == target;        
    }

    if runops(nums, inx + 1, target, acc * nums[inx], do2 ) {
        return true;
    }

    if runops(nums, inx + 1, target, acc + nums[inx], do2 ) {
        return true;
    }
    if do2 {
        let new_acc:usize = (acc.to_string() + &nums[inx].to_string())
            .parse()
            .unwrap();

        if runops(nums, inx + 1, target, new_acc, do2) {
            return true;
        }
    }

    false
}

fn day7(input: &str) -> (usize, usize) {
    let mut r1 = 0;
    let mut r2 = 0;
    for line in input.lines() {
        let (target,numstr) = line
            .split_once(":")
            .map(|(f,s)| (f.parse::<usize>().unwrap(), s) )
            .unwrap();
        let nums = numstr
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap())
            .collect::<Vec<_>>();

        if runops(&nums, 1, target, nums[0], false ) {
            r1 += target;
        }
        if runops(&nums, 1, target, nums[0], true ) {
            r2 += target;
        }
    }
    
    (r1, r2)
}

fn main() {
    let (p1, p2) = day7(INPUT);
    println!("{}\n{}", p1, p2);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1() {
        let input = include_str!("inputs/day7.test1.txt");
        let (p1, p2) = day7(input);

        assert_eq!(p1, 3749);
        assert_eq!(p2, 11387);
    }

    #[test]
    fn test_main() {
        let (p1, p2) = day7(INPUT);

        assert_eq!(p1, 12940396350192);
        assert_eq!(p2, 106016735664498);
    }
}

File: python/day7.py
# Here I started by making a binary number as the two "ops" 0 => + and 1 => *
# Then the part2 cureball was adding another op of || so I had to find a way
# to make tenary numbers - but this was slow
# It's also somewhat horrible and I wish I'd spent longer thinking about it first!

def runops(nums, ops, target):
    acc = nums[0]
    for i,op in enumerate(ops):
        if op == '2':
            acc = int(str(acc) + str(nums[i+1]))
        elif op == '1':
            acc *= nums[i+1]
        elif op == '0':
            acc += nums[i+1]
        if acc > target:
            return False
    return acc == target

# Originally this wasn't memoized and it turns
# out this is slow. Memoizing improved it a bit!
# (see also day7_faster)
memoize = {}
def ternary(num, l):
    n = num
    if (num,l) in memoize:
        return memoize[(num,l)]
    digits = [0]*l
    for i in range(l):
        (num,r) = divmod(num,3)
        digits[i] = r
    res = ''.join(str(x) for x in reversed(digits))
    memoize[(n,l)] = res
    return res

def day7(input):
    result1 = 0
    result2 = 0
    lc = 0
    for line in input:
        print(f"{lc}/{len(input)}")
        lc += 1
        l = line.split()
        target = int(l[0][:-1])
        nums = [int(i) for i in l[1:]]

        oppow = 2 ** (len(nums) - 1)
        fo = f"0{len(nums) - 1}b"
        for i in range(oppow):
            ops = format(i, fo)
            if runops(nums,ops,target):
                result1 += target
                break

        oppow = 3 ** (len(nums) - 1)
        for i in range(oppow):
            ops = ternary(i, len(nums) - 1)
            if runops(nums,ops,target):
                result2 += target
                break

    print(f"{result1} {result2}")

#3749 11387
input = open("src/bin/inputs/day7.test1.txt", "r").readlines()
day7(input)

#12940396350192 106016735664498
input = open("src/bin/inputs/day7.txt", "r").readlines()
day7(input)